pilot/pkg/bootstrap/server.go

```diff
@@ -792,26 +792,37 @@ func (s *Server) addTerminatingStartFunc(fn server.Component) {
 }
 
 func (s *Server) waitForCacheSync(stop <-chan struct{}) bool {
+	start := time.Now()
+	log.Info("Waiting for caches to be synced")
 	if !cache.WaitForCacheSync(stop, s.cachesSynced) {
 		log.Errorf("Failed waiting for cache sync")
 		return false
 	}
+	log.Infof("All controller caches have been synced up in %v", time.Since(start))
+
 	// At this point, we know that all update events of the initial state-of-the-world have been
-	// received. Capture how many updates there are
-	expected := s.XDSServer.InboundUpdates.Load()
-	// Now, we wait to ensure we have committed at least this many updates. This avoids a race
+	// received.We wait to ensure we have committed at least this many updates. This avoids a race
 	// condition where we are marked ready prior to updating the push context, leading to incomplete
 	// pushes.
-	if !cache.WaitForCacheSync(stop, func() bool {
-		return s.XDSServer.CommittedUpdates.Load() >= expected
-	}) {
+	if !cache.WaitForCacheSync(stop, s.pushContextReady) {
 		log.Errorf("Failed waiting for push context initialization")
 		return false
 	}
 
 	return true
 }
 
+// pushContextReady indicates whether pushcontext has processed all inbound config updates.
+func (s *Server) pushContextReady() bool {
+	expected := s.XDSServer.InboundUpdates.Load()
+	committed := s.XDSServer.CommittedUpdates.Load()
+	if committed < expected {
```

Comment:
I think this is not right

Reply:
Why it is not correct?

Reply:
expected needs to be fixed. We should not update itevery time, otherwise its a constantly moving target. We should set it once after caches are synced, then we will check we process at least all of those updates (ie initial state, but maybe not a new resource created after pilot started)

Reply:
I see. Thanks I did not get it..